pkg/operator/controller/sync-http-error-code-configmap/controller.go

```diff
@@ -0,0 +1,247 @@
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	runtimecontroller "sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "error_page_configmap_controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+// New creates a new controller that syncs HTTP error page configmaps between
+// namespaces.
+func New(mgr manager.Manager, config Config) (runtimecontroller.Controller, error) {
+	operatorCache := mgr.GetCache()
+	reconciler := &reconciler{
+		cache:    operatorCache,
+		client:   mgr.GetClient(),
+		config:   config,
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := runtimecontroller.New(controllerName, mgr, runtimecontroller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// If the ingresscontroller's error-page configmap reference changes,
+	// reconcile the ingresscontroller.
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc:  func(e event.CreateEvent) bool { return reconciler.hasConfigMap(e.Object) },
+		DeleteFunc:  func(e event.DeleteEvent) bool { return reconciler.hasConfigMap(e.Object) },
+		UpdateFunc:  func(e event.UpdateEvent) bool { return reconciler.configMapChanged(e.ObjectOld, e.ObjectNew) },
+		GenericFunc: func(e event.GenericEvent) bool { return reconciler.hasConfigMap(e.Object) },
+	}); err != nil {
+		return nil, err
+	}
+
+	// Index ingresscontrollers by spec.httpErrorCodePages.name so that
+	// configmapToIngressController and configmapIsInUse can look up
+	// ingresscontrollers that reference the configmap.
+	if err := mgr.GetFieldIndexer().IndexField(
+		context.Background(),
+		&operatorv1.IngressController{},
+		"spec.httpErrorCodePages.name",
+		client.IndexerFunc(func(o client.Object) []string {
+			ic := o.(*operatorv1.IngressController)
+			if len(ic.Spec.HttpErrorCodePages.Name) == 0 {
+				return []string{}
+			}
+			return []string{ic.Spec.HttpErrorCodePages.Name}
+		})); err != nil {
+		return nil, fmt.Errorf("failed to create index for ingresscontroller: %w", err)
+	}
+
+	isInNamespace := func(namespace string) func(o client.Object) bool {
+		return func(o client.Object) bool {
+			log.Info("isInNamespace", "name", o.GetName(), "namespace", o.GetNamespace(), "result", o.GetNamespace() == namespace)
```

Comment:
Can we please remove this log statement to avoid log spam? Thanks!

Reply:
done